src/Raven.Server/Documents/Queries/Graph/EdgeQueryStep.cs

```diff
@@ -85,6 +96,12 @@ private async Task InitializeRightAsync(ValueTask rightTask)
         private async Task InitializeLeftAsync(ValueTask leftTask)
         {
             await leftTask;
+            //At this point we know we are not going to yield results we can skip running right hand side
+            if (Left.IsEmpty())
+            {
+                _index = 0;
+                return;
+            }
             var rightTask = _right.Initialize();
```

Comment:
Shouldn't you do the check on the right side as well? You know if there is no one there, the whole thing will return no results.

Reply:
The majority of the performance is lost in the initialization so the performance gain to check afterward is much smaller since the logic is mostly in memory

Reply:
done

Reply:
We actually do a fair amount of work in edge steps and recursive steps during the actual `CompleteInitialization` that I want to save

Reply:
is there anything else i missed?